Before you turn this homework in, make sure everything runs as expected. First, **restart the kernel** (in the menubar, select Kernel$\rightarrow$Restart) and then **run all cells** (in the menubar, select Cell$\rightarrow$Run All). If you get error messages, the grader will get the same errors and won't be able to grade your work. 

Make sure you fill in any place that says `YOUR CODE HERE` or "YOUR ANSWER HERE". You may delete the line(s) that say "**raise** NotImplementedError()" and replace them with your actual work. Write all your work in the cells provided in this notebook. **DO NOT** add or delete any cells. 

Please fill in your name in the cell below.

In [1]:
NAME = "Jennifer Luu"

---

# Homework 4

In all problems below, please comment your code sufficiently well so that the grader can follow what you are doing with ease. For non-coding answers, please make sure to formulate your explanation and answers in the form of complete English sentences. It is not sufficient to just leave comments in the code (without a full sentence explanations in problems that ask for those) or vice versa (to leave full sentence explanations but not code comments). You will need both to get full credit. 

## Problem 1

Write a function named "border" that takes as its input two integers n and m ($1 \le n,m \le 100$) and outputs a NumPy array of size $(n,m)$. Your array should be filled with zeros except for the "border" (that is, the first and last column and the first and last row) which should be filled with ones. 

In [2]:
import numpy as np

def border(n, m):
        
    """Creates an array with shape (n, m) that is all zeros
    except for the border (i.e., the first and last rows and
    columns), which should be filled with ones."""
    # YOUR CODE HERE
    if type(n)==int and type(m)==int and 1<=n<=100 and 1<=m<=100: #condition check 
        my_array = np.zeros((n,m), int) #make an array full of zeros 
        #find the slices for the first/last row/column and generate arrays of 1 to fill
        my_array[0,:] = [1]*n   #top row
        my_array[-1,:] = [1]*n  #bottom row
        my_array[:,0] = [1]*m   #first column
        my_array[:,-1] =[1]*m   #last column
        return my_array
    else:
        return false

#np.array([[1, 1, 1], [1, 0, 1], [1, 1, 1]])

# check that your function works as expected (if it does this check should return "True")
np.array_equal(border(3,3), np.array([[1, 1, 1], [1, 0, 1], [1, 1, 1]]))

True

## Problem 2

(a) Suppose that x and y are both NumPy arrays of size $(1,n)$. 
Write a function that computes Pearson's sample correlation coefficient $r$ 
between the entries in x and y. 

Recall, that Pearson's sample correlation coefficient is defined as 
$$r(x,y) = \frac{\sum_{i=1}^n (x_i - \bar{x})(y_i - \bar{y})}{\sqrt{\sum\limits_{i=1}^n (x_i - \bar{x})^2 \sum\limits_{i=1}^n(y_i - \bar{y})^2}} $$

In [14]:
import math 

def pearsons_correlation(x, y):
    """Computes Pearson's correlation coefficient between vectors x and y."""
    
    # YOUR CODE HERE
    x_bar = np.mean(x) #find the mean of x and y
    y_bar = np.mean(y)
    
    x_dev = [entry-x_bar for entry in x] #get deviations for each element in x and y 
    y_dev = [entry-y_bar for entry in y]
    
    top = np.sum((np.multiply(x_dev, y_dev))) #perform element-wise multiplication and sum 
    
    xdev_sqr = np.sum([entry**2 for entry in x_dev]) #square the deviations and sum in x and y 
    ydev_sqr = np.sum([entry**2 for entry in y_dev])

    btm_sqr = np.sqrt(xdev_sqr*ydev_sqr) #take square root of the product of the sums
    
    pears_cor = top/btm_sqr #compute r 
    
    return pears_cor
    
# check that your function works
# if it does, the comparison below should return "True"

import numpy as np

np.random.seed(10)
n = 10
x = np.random.random((1,10))
y = np.random.random((1,10))

round(pearsons_correlation(x,y),6) == round(np.corrcoef(x,y)[0,1],6)

#print(pearsons_correlation(x,y))
#print(round(np.corrcoef(x,y)[0,1],6))
# we're rounding, because the built-in NumPy method for correlation 
# may use more efficient arithmetic than we do in our home-made function. 

True

(b) Create a NumPy array filled with uniform(0,1) random values with n rows and m columns (for the combinations of n and m listed below). In each case, find the maximum of the Pearson correlation coefficients for each of the $\binom{m}{2}$ pairs of columns in your array. 

That is, if the columns of your array are $x_1, \ldots, x_m$, find 
$$\max(r) = \max\limits_{a,b \in \{1,\ldots,m\}, a \neq b} r(x_a, x_b)$$

Enter the values you find into the table below (rounded to two after decimal digits).

In [15]:
import numpy as np
np.random.seed(10)

def max_corr(n,m):
    array = np.random.random((n,m)) #a numpy array with uniform(0,1) r.v with n rows and m columns 
    """find maximum correlation between columns of (n,m) random uniform (0,1) array"""
    # YOUR CODE HERE
    r = [] #initialize empty list to hold all r's
    for i in range(m):
        for j in range(m):
            if i==j:
                continue
            else: 
                x = array[:,i]
                y = array[:,j]
                corr = pearsons_correlation(x, y)
                r.append(corr)
    return max(r)

max_corr(100,100)


    

0.41454717370423394

## |  rows n | columns m  | max(r)  |
|:-:|:-:|:-:|
| 10  | 20  | |
| 10  | 50  |  |
| 10  | 100  |  |
| 100  | 20  |  |
| 100  | 50  |  |
| 100  | 100  |  |

Copy and paste this table to the Markdown cell below and complete it. 

## |  rows n | columns m  | max(r)  |
|:-:|:-:|:-:|
| 10  | 20  | 0.73 |
| 10  | 50  | 0.82 |
| 10  | 100  | 0.96 |
| 100  | 20  | 0.26 |
| 100  | 50  | 0.34 |
| 100  | 100  | 0.41 |

Copy and paste this table to the Markdown cell below and complete it. 

## Problem 3

Suppose that $X$ is a matrix with n rows and m columns. In statistics, these matrices arise freqently when we collect (numerical) data on m variables which are each observed on n independent individuals. Suppose we denote the columns of $X = (x_1, \ldots, x_m)$. Then the covariance matrix of $X$ is defined as 

$$ \mbox{Cov}(X) = \left( \begin{array}{cccc}
Var(x_1) & Cov(x_1, x_2) & \cdots & Cov(x_1, x_m) \\
Cov(x_2, x_1) & Var(x_2) & \cdots & Cov(x_2, x_m) \\
\vdots & \vdots & \ddots & \vdots \\
Cov(x_m, x_1) & Cov(x_m, x_2) & \cdots & Var(x_m) 
\end{array} \right) $$

where Var($x_i$) is the sample variance of the entries in the $i^{th}$ column of $X$ and Cov($x_i, x_j$) ($i \neq j$) is the sample covariance of the entries in columns $i$ and $j$. 

$$ Cov(x_i, x_j) = \frac{1}{n-1}\sum\limits_{k=1}^n (x_{ki} - \bar{x}_i)(x_{kj}-\bar{x_j})$$

That is, the covariance matrix is a square $m\times m$ matrix whose diagonal entries are the sample variances of the columns of $X$ and whose off-diagonal entries are covariances between two columns of $X$, respectively. 

Write *your own* function that takes a Numpy array of shape (n,m) as its input and returns the  Covariance matrix of the array. Do not use any built-in functions to compute variance and/or covariance. Write your own functions, instead. It's ok to use NumPy routines for sums, squares, or square-roots. 

Check your work by comparing your result to that of the ```np.cov()``` function applied to the same input.

In [5]:
def covariance_matrix(X):
    """ Finds the covariance matrix of (n,m) shaped array X"""
    # YOUR CODE HERE
    columns = len(X[0]) #number of columns
    rows = len(X) #number of rows
    cov_matrx = np.zeros((columns, columns)) #initialize empty m x m array 
    scale = 1/(rows-1)
    
    for i in range(columns): #iterate through each position in the cov matrix 
        for j in range(columns):
            xi_bar = np.mean(X[:,i]) #obtain means for xi and xj
            xj_bar = np.mean(X[:,j])
            
            prod1 = [(xki - xi_bar) for xki in X[:,i]] #compute the differences for each column
            prod2 = [(xkj - xj_bar) for xkj in X[:,j]]
            
            prod_list = np.multiply(prod1,prod2) #perform element-wise multiplication 
            
            cov_matrx[i,j] = (sum(prod_list))*scale #sum and multiply by 1/(n-1)
            
    return cov_matrx
    
## check your work
X = np.random.random((5,3))  # or make up some other matrix for X 
np.array_equal(np.round(np.cov(X.T),4), np.round(covariance_matrix(X),4)) 
# again, we're rounding because your "hand" computation will differ slightly from NumPy's internal computations


True

## Problem 4

The code given below (don't change the seed, please) generates n=10,000 IID random samples from Student's t-distribution with degree of freedom $\nu = 5$. Let $X \sim t(\nu=5)$ be a random variable with a $t_5$ distribution. For each of the following quantities, find their values using the methods your have learned in Math 161A. State your answer together with a short description of how you find it. Also find the best estimate of each quantity you can produce based on the 10,000 generated random numbers.

In [6]:
from scipy import stats
import numpy as np

np.random.seed(10)
Data = np.array(stats.t.rvs(df=5,size=10000))

(a) $P(-1 \le X < 2)$

In [7]:
# YOUR CODE HERE
successes = (-1<=Data)&(Data<2)
prob = sum(successes)/len(Data)
print(prob)

0.7648


The probability that -1<=X<2 is 0.7648. 

(b) The $77^{th}$ percentile of $X$.

In [8]:
# YOUR CODE HERE
import math as math

sorted_data = np.sort(Data) #sort from smallest to largest number 

index = len(Data)*0.77 #multiply by the length of the array by 77 
 
if (index%1 >= 5): #check decimal to take either integer ceiling or floor for index
    index = int(math.ceiling(index))
elif (index%1 <5):
    index = int(math.floor(index))
    
per = sorted_data[index] #obtain 77th percentile by indexing in sorted array

print(per)

0.796361368761396


The 77th percentile of Data is 0.79636.

(c) Var($X$) (that is the sample variance of $X$)

In [9]:
# YOUR CODE HERE
x_bar = np.mean(Data) #find sample mean 
sqrd_dev = sum([(entry-x_bar)**2 for entry in Data]) #find squared deviations 
btm = len(Data)-1 #bottom term 
sample_var = sqrd_dev/btm

print(sample_var)


1.6499633760480088


The sample variance of X is 1.64996.

## Problem 5

The code below generates a NumPy array X of shape (12,8) filled with integers. Please don't change the seed.

In [10]:
import numpy as np

np.random.seed(10)
X = np.random.randint(10,size = (12,8)) 
X

array([[9, 4, 0, 1, 9, 0, 1, 8],
       [9, 0, 8, 6, 4, 3, 0, 4],
       [6, 8, 1, 8, 4, 1, 3, 6],
       [5, 3, 9, 6, 9, 1, 9, 4],
       [2, 6, 7, 8, 8, 9, 2, 0],
       [6, 7, 8, 1, 7, 1, 4, 0],
       [8, 5, 4, 7, 8, 8, 2, 6],
       [2, 8, 8, 6, 6, 5, 6, 0],
       [0, 6, 9, 1, 8, 9, 1, 2],
       [8, 9, 9, 5, 0, 2, 7, 3],
       [0, 4, 2, 0, 3, 3, 1, 2],
       [5, 9, 0, 1, 0, 1, 9, 0]])

(a) Write a function called "swap" that takes as its input the array $X$ and two integers n and m ($ 1 \le n,m \le 12, n \neq m$) and returns the matrix X in which rows n and m have been swapped. 

In [11]:
def swap(X,n,m):
    """swaps rows n and m of array X"""
    # YOUR CODE HERE
    if 1<=n<=12 and 1<=m<=12 and n!=m: #conditon check 
        n=n-1 #decrement the integers by 1 to accomodate for python indexing
        m=m-1

        Y = X.copy() #create an independent copy of X
        row_n = Y[n,:] #obtain row n
        row_m = Y[m,:] #obtain row m

        X[m,:] = row_n #swap with real array X
        X[n,:] = row_m

        return X
    else:
        return False
    
    
# check your work
np.array_equal(swap(X,1,2), np.array([[9, 0, 8, 6, 4, 3, 0, 4],
       [9, 4, 0, 1, 9, 0, 1, 8],
       [6, 8, 1, 8, 4, 1, 3, 6],
       [5, 3, 9, 6, 9, 1, 9, 4],
       [2, 6, 7, 8, 8, 9, 2, 0],
       [6, 7, 8, 1, 7, 1, 4, 0],
       [8, 5, 4, 7, 8, 8, 2, 6],
       [2, 8, 8, 6, 6, 5, 6, 0],
       [0, 6, 9, 1, 8, 9, 1, 2],
       [8, 9, 9, 5, 0, 2, 7, 3],
       [0, 4, 2, 0, 3, 3, 1, 2],
       [5, 9, 0, 1, 0, 1, 9, 0]]))  

True

(b) Write a function called "sort_by_column" that takes as its input the array X and an integer k ($1\le k \le 8$) and sorts the rows of X by the values of column k. Numbers that were in the same row before, should still be in the same row. Don't worry about ties - if there are ties, I don't care how you order them. 

In [12]:
def sort_by_column(X,k):
    """ sort rows of X by values in column k"""
    # YOUR CODE HERE
    raise NotImplementedError()